Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceToTemplateInterface.h

```diff
@@ -48,8 +48,9 @@ namespace AzToolsFramework
             //! @param templateId The id of the template to update.
             //! @param instanceToExclude An optional reference to an instance of the template being updated that should not be refreshes as part of propagation.
             //!     Defaults to nullopt, which means that all instances will be refreshed.
+            //! @param immediate An optional flag whether to apply the patch immediately (needed for Undo/Redos) or wait until next system tick.
             //! @return True if the template was patched correctly, false if the operation failed.
-            virtual bool PatchTemplate(PrefabDomValue& providedPatch, TemplateId templateId, InstanceOptionalReference instanceToExclude = AZStd::nullopt) = 0;
+            virtual bool PatchTemplate(PrefabDomValue& providedPatch, TemplateId templateId, InstanceOptionalReference instanceToExclude = AZStd::nullopt, bool immediate = false) = 0;
```

Comment:
Generally this change looks good, but it does have me concerned a bit. Prefabs have been designed to support multi-threading in the original design. So far the performance has exceeded the original Slice performance, so any major performance work like executing the multi-threaded part of the design haven't had the need to be prioritized over more pressing feature work. However changes like this that require immediate processing become a headache when adding multi-threading and also hint at a potentially larger underlying problem with the designed threading model.
It seems that this might be down to systems interacting with entities that then needs to be reflected in the prefab that created those entities. Is that correct? If so, that could mean that entities are not sufficiently isolated from their entities which may come up as well when the new DOM Property Editor is used. Do you think there's a way to fix this by deferring calls from the entities as well or is does too much depend on immediate updates?

Reply:
Undo system does not support async callbacks, I have looked into that but it's not trivial to add. I think part of the problem is that we're still trying to support slices which rely on same interfaces and use same systems. So trying to refactor something for prefabs propagates to having to make exponentially more and more changes and potentially creating bugs in slices. I see a lot of problems with prefab code being caused by trying to conform to using original slices API and undo system API. 

I don't know good solution here outside of re-writing much of prefab code with new API and new undo system. Otherwise we're going to keep adding more hacks and flags and unnecessary functions just to get things working.

Reply:
That's good feedback, thanks for looking into that. Sounds like detaching Prefabs from the legacy Slice code is becoming increasingly important. Hopefully the move to the proposed Dom Property Editor will be a big step forward in that regard. Note though that with the Dom Property Editor we also want updating the entities in the viewport to reflect the changes to the Prefabs to be time-sliced/multi-threaded, so it's worth keeping asynchronous operations in mind in genera.